README.md

This package provides an easy way to integrate CapnoTrainer GO device with your Flutter app. The basic idea is to use *flutter_blue_plus* package, scan for GO devices and pass the *BluetoothDevice* object and a callback function to the CapnoTrainer connect method. Once done, call disconnect on the same CapnoTrainer object. The data is received on callback handler. 

## Features

1) Easier way to connect to GO device.
2) Different data is already parsed and ready to be further processed or present. 
3) BLE handling is done by the package so no need to worry about managing it separately. 

## Getting started

See the examples folder. 

*P.S: The example is tested on Android phone. However, it will work for iOS as well as long as relevant permissions are handled under ios folder.*

## Usage

Full example is available in example folder



CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.
1) Added a CapnoTrainer class 
2) Added an example that shows how to use CapnoTrainer Flutter SDK including: 
   1) connect method (pass a callback and GO device Bluetooth object)
   2) how to handle the data and different status codes in callback 
   3) how to disconnect 
   4) basic BLE scanning

3) The callback function uses C-like status codes. Enums are in utils for refences. 



examples/lib/main.dart
import 'package:flutter/material.dart';
import 'package:examples/widgets.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter CapnoTrainer SDK',
      theme: ThemeData( primarySwatch: Colors.purple,  ),
      home: const MyHomePage(title: 'Flutter CapnoTrainer SDK'),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  void _refresh() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),

      ),
      body: Center(
        child: BLEScreen(),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _refresh,
        tooltip: 'Refresh',
        child: const Icon(Icons.refresh),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/capnotrainer_sdk.dart
import 'package:capnotrainer_sdk/capnotrainer_utils.dart';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';
import 'dart:async';
import 'package:capnotrainer_sdk/capnotrainer_go_parser.dart';

class CapnoTrainer {

  final bool _printLog = false;
  late BluetoothDevice device;
  late Function onDataReceived;
  late BluetoothCharacteristic _readCharacteristic;
  late BluetoothCharacteristic _writeCharacteristic;

  // device identifier
  Guid capnoServiceGuid = Guid("00001000-0000-1000-8000-00805f9b34fb");
  Guid capnoReadCharGuid =  Guid("00001002-0000-1000-8000-00805f9b34fb");
  Guid capnoWriteCharGuid =  Guid("00001001-0000-1000-8000-00805f9b34fb");

  // local state
  bool _deviceBleState = false;
  BluetoothConnectionState _deviceConnectionState = BluetoothConnectionState.disconnected;
  late StreamSubscription _deviceStateSubscription;
  late StreamSubscription _deviceReadSubscription;
  Timer _deviceWriteTimer = Timer(const Duration(hours: 1), () => {},);
  late CapnoTrainerGoParser _parser;

  bool get isConnected => _deviceConnectionState == BluetoothConnectionState.connected;

  Future<void> connect(BluetoothDevice d, Function onData) async {
    device = d;
    onDataReceived = onData;

    if (!isConnected){
      try{
        await device.connect(autoConnect: true);
      } catch(e){
        if (_printLog) print ("[CapnoTrainer SDK] ${e}");
      }
      _deviceStateSubscription = device.connectionState.listen(
          _handleDeviceState,
          onError: (e)=>{ if (_printLog) print("[CapmoTrainer SDK] ${e}")},
          onDone: () => { if (_printLog) print ("[CapnoTrainer SDK] ") }
      );
    }
  }

  Future<void> disconnect() async {
    if (_deviceWriteTimer.isActive) _deviceWriteTimer.cancel();
    await _deviceStateSubscription.cancel();
    await _deviceReadSubscription.cancel();
    await device.disconnect(timeout: 30);
    onDataReceived(<CapnoTrainerDataPoint>[], CapnoTraienrStatusCodes.CODE_DISCONNECTED);
    _deviceConnectionState = BluetoothConnectionState.disconnected;
  }

  _handleDeviceState(BluetoothConnectionState state) async {
    _deviceConnectionState = state;
    _deviceBleState = BluetoothConnectionState.connected == state;
    if (state == BluetoothConnectionState.connected){
      await _handleSetupDevice();
      onDataReceived(<CapnoTrainerDataPoint>[], CapnoTraienrStatusCodes.CODE_CONNECTED);
    }
  }

  Future<void> _handleSetupDevice() async {
    List<BluetoothService> services = await device.discoverServices();
    for (BluetoothService service in services) {
      if (service.uuid == capnoServiceGuid) {
        for (BluetoothCharacteristic characteristic
        in service.characteristics) {
          if (characteristic.uuid == capnoReadCharGuid) {
            _readCharacteristic = characteristic;
          } else if (characteristic.uuid == capnoWriteCharGuid) {
            _writeCharacteristic = characteristic;
          }
        }
      }
    }

    try{
      await _readCharacteristic.setNotifyValue(true);
    } catch(e){
      if (_printLog) print ("[CapnoTrainer SDK] ${e}");
    }

    await Future.delayed(const Duration(milliseconds: 100));

    _deviceReadSubscription = _readCharacteristic.lastValueStream.listen(
      _onDeviceData, onError: _onDeviceError, onDone: _onDeviceDone
    );

    _parser = CapnoTrainerGoParser(_handleDeviceWrite, onDataReceived);

    if (_deviceWriteTimer.isActive) _deviceWriteTimer.cancel();
    _deviceWriteTimer = Timer.periodic(const Duration(milliseconds: 100), _onDeviceWrite );

  }

  _onDeviceData(List<int> data) async {
    await _parser.handlePackagedDataRead(data);
  }

  _onDeviceError(e){
    if (_printLog) print ("[CapnoTrainer SDK] ${e}");
  }

  _onDeviceDone(){

  }

  _onDeviceWrite(Timer timer) async {
    if (!_parser.isReceivedPressure) {
      await _parser.handleBarometricPressureRequest();
    } else if (!_parser.isReceivedTemperature){
      await _parser.handleGasTemperatureRequest();
    } else {
      await _parser.handleBatteryStatusRequest();
      await _parser.handleStartStreamRequest();
      timer.cancel();
    }
  }

  _handleDeviceWrite(List<int> cmd) async {
    try{
      if (_deviceBleState){
        await _writeCharacteristic.write(cmd);
      } else {
        if (_printLog) print ("[CapnoTrainer SDK] Device is not connected. Unable to write to characteristics");
      }
    } catch (e) {
      if (_printLog) print ("[CapnoTrainer SDK] ${e}");
    }
  }

}

lib/capnotrainer_go_parser.dart
import 'dart:math';
import 'package:capnotrainer_sdk/capnotrainer_utils.dart';

class CapnoTrainerGoParser{

  final Function _handleSendCmd;
  final Function _onDataReceived;
  final bool _printLog = false;

  final List<CapnoTrainerDataPoint> _dataCo2 = [];
  int _samplesCo2 = 0;
  final int _rateCo2 = 100;

  // basic parsing holder
  List<int> _incomingDataLast = [];

  // write variables
  bool _isReceivedTemperature = false;
  bool _isReceivedPressure = false;
  bool _isReceivedBattery = false;

  void set setDataRawCo2(CapnoTrainerDataPoint point) => _dataCo2.add(point);
  void set setSampleCo2(int samples) => _samplesCo2 = samples;

  bool get isReceivedTemperature => _isReceivedTemperature;
  bool get isReceivedPressure => _isReceivedPressure;
  bool get isReceivedBattery => _isReceivedBattery;
  int get samplesCo2 => _samplesCo2;
  int get rateCo2 => _rateCo2;

  CapnoTrainerGoParser(this._handleSendCmd, this._onDataReceived);

  handlePackagedDataRead(List<int> incoming){

    List<int> data = [..._incomingDataLast, ...incoming];
    _incomingDataLast = [];
    int count = 0;

    for (count = 0; count < data.length; count++) {
      bool isCmdByte = data.elementAt(count) >= 0x80;
      if (isCmdByte) {
        int subDataSize = data.elementAt(1) + 2;
        if (data.length >= count + subDataSize) {
          List<int> subData =
          data.getRange(count, count + subDataSize).toList();
          if (handleCorrectChecksum(subData) && subData.length == subDataSize) {
            handleCmdByteParsing(subData);
          } else {

          }
        } else {
          _incomingDataLast = data.getRange(count, data.length).toList();
        }
      }
    }
  }

  handlePackageDataWrite(List<int> cmd) async {
    cmd = _handleAddChecksum(cmd);
    if (handleCorrectChecksum(cmd)){
      await _handleSendCmd(cmd);
    }
  }

  bool handleCorrectChecksum(List<int> cmd) {
    int sum = 0x00;
    int checksum = 0x00;

    if (cmd.elementAt(0) < 0x80) {
      return false;
    } else {
      int size = cmd.elementAt(1) + 2;
      if (size != cmd.length) {
        return false;
      }
      for (int i = 0; i < size; i++) {
        sum = sum + cmd.elementAt(i);
      }
      checksum = (sum & 0x7F + checksum);
      checksum &= 0x7F;

      if (checksum == 0x00)
        return true;
      else
        return false;
    }
  }


  void handleCmdByteParsing(List<int> data) {
    int cmdByte = data.elementAt(0);
    switch (cmdByte) {
      case 0x80:
        this._handleStartStreamResponse(data);
        break;
      case 0xC9:
        this._handleStopStreamResponse(data);
        break;
      case 0x84:
        int isbType = data.elementAt(2);
        switch (isbType) {
          case 0:
            break;
          case 1:
            _handleBarometricPressureResponse(data);
            break;
          case 4:
            _handleGasTemperatureResponse(data);
            break;
          default:
          // TODO: Add more response for ISB types.
            break;
        }
        break;

      case 0xFB:
        int infoType = data.elementAt(2);
        switch (infoType) {
          case 0x07:
            _handleBatteryStatusResponse(data);
            break;
          default:
          // TODO: Add more response for infor types.
            break;
        }
        break;
      case 0xC8:
        _handleNackErrorCodes(data);
        break;
      default:
      // print ("CapnoTrainer 7.0 - Unknown data type received");
    }
  }

  List<int> _handleAddChecksum(List<int> cmd) {
    int checksum = 0x00;
    int size = cmd.elementAt(1) + 1;
    for (int i = 0; i < size; i++) {
      checksum = checksum + cmd.elementAt(i);
    }
    checksum = (~checksum + 1) & 0x7F;
    cmd[cmd.length - 1] = checksum;

    return cmd;
  }

  String _handleNackErrorCodes(List<int> cmd) {
    int errorCode = cmd.elementAt(2);
    switch (errorCode) {
      case 0x00:
        return "BOOTCODE_WAITE";
        break;
      case 0x01:
        return "INVALID_COMMAND";
        break;
      case 0x02:
        return "CHECKSUM_ERROR";
        break;
      case 0x03:
        return "TIMEOUT_ERROR";
        break;
      case 0x04:
        return "INVALID_BYTE_COUNT";
        break;
      case 0x05:
        return "INVALID_BYTE_DATA";
        break;
      case 0x06:
        return "SYSTEM_FAULTY";
        break;
      default:
        return "";
        break;
    }
  }


  Future<void> _handleStopStreamRequest() async {
    List<int> cmd = [0xC9, 0x01, 0x00];
    await handlePackageDataWrite(cmd);
  }

  void _handleStopStreamResponse(List<int> data) {
    // stop stream
  }

  Future<void> handleStartStreamRequest() async {
    List<int> cmd = [0x80, 0x02, 0x00, 0x00];
    await handlePackageDataWrite(cmd);
  }

  void _handleStartStreamResponse(List<int> data) {

    try{
      int nbf = data.elementAt(1);
      int sync = data.elementAt(2);
      int co2wb1 = data.elementAt(3);
      int co2wb2 = data.elementAt(4);
      double co2 = (((128.0 * co2wb1.toDouble()) + co2wb2.toDouble()) - 1000.0) / 100.0;
      double x = samplesCo2 * 1000 / rateCo2;

      setDataRawCo2 = CapnoTrainerDataPoint(x, co2);
      setSampleCo2 = samplesCo2 + 1;

      if ( _dataCo2.length > 10 ) {
        List<CapnoTrainerDataPoint> co2Data = _dataCo2;
        _onDataReceived(co2Data, CapnoTraienrStatusCodes.CODE_RAW_CO2_DATA);
        _dataCo2.clear();
      }

      // send battery update request every 5 minutes.
      if ( x ~/ 1000 % 300 == 0){
        handleBarometricPressureRequest();
      }

      if ( nbf > 6 ) {
        int dpi = data.elementAt(5);
        if ( dpi == 1 ) {
          int dpb1 = data.elementAt(6);
          int dpb2 = data.elementAt(7);
          int dpb3 = data.elementAt(8);
          int dpb4 = data.elementAt(9);
          int dpb5 = data.elementAt(10);
          if ( dpb1 & 0x10 == 0x10 && dpb4 & 0x04 == 0x04 ){
            _onDataReceived(<CapnoTrainerDataPoint>[], CapnoTraienrStatusCodes.CODE_AIRWAY_BLOCKED);
          }

          bool breathDetected = dpb1 & 0x40 == 0x40;
          bool canNotZero = dpb1 & 0x10 == 0x10;
          // if ( !breathDetected && !canNotZero ){
          //   _onDataReceived([], CapnoTraienrStatusCodes.CODE_READY_TO_ZERO);
          // }

        } else if ( dpi == 2  ){
          int dpb1 = data.elementAt(6);
          int dpb2 = data.elementAt(7);
          double etco2 = (dpb1 * 128 + dpb2).toDouble() / 10.0;
          _onDataReceived(
              [CapnoTrainerDataPoint(samplesCo2*1000/rateCo2, etco2)],
              CapnoTraienrStatusCodes.CODE_ETCO2_DATA
          );
        } else if ( dpi == 3 ){
          int dpb1 = data.elementAt(6);
          int dpb2 = data.elementAt(7);
          double bpm = (dpb1 * 128 + dpb2).toDouble();
          _onDataReceived(
              [CapnoTrainerDataPoint(samplesCo2*1000/rateCo2, bpm)],
              CapnoTraienrStatusCodes.CODE_BPM_DATA
          );
        } else if ( dpi == 4 ){
          int dpb1 = data.elementAt(6);
          int dpb2 = data.elementAt(7);
          double insp = (dpb1 * 128 + dpb2).toDouble() / 10.0;
          _onDataReceived(
              [CapnoTrainerDataPoint(samplesCo2*1000/rateCo2, insp)],
              CapnoTraienrStatusCodes.CODE_INSP_CO2_DATA
          );
        }
      }

    } catch(e){
      if (_printLog) print("[CapnoTrainer SDK] $e}");
    }
  }

  Future<void> handleBarometricPressureRequest() async {
    List<int> cmd = [0x84, 0x02, 0x01, 0x00];
    await handlePackageDataWrite(cmd);
  }

  void _handleBarometricPressureResponse(List<int> data) {
    double _pressure = (data.elementAt(3) * 128 + data.elementAt(4)).toDouble();
    if (_printLog) print ("[CapnoTrainer SDK] CapnoTrainer Go pressure is $_pressure");
    _isReceivedPressure = true;
  }

  Future<void> handleGasTemperatureRequest() async {
    List<int> cmd = [0x84, 0x02, 0x04, 0x00];
    await handlePackageDataWrite(cmd);
  }

  void _handleGasTemperatureResponse(List<int> data) {
    double _temperature =
    ((data.elementAt(3) * 128 + data.elementAt(4)) / 2).toDouble();
    if (_printLog) print ("[CapnoTrainer SDK] CapnoTrainer Go temperature is $_temperature");
    _isReceivedTemperature = true;
  }

  Future<void> handleBatteryStatusRequest() async {
    List<int> cmd = [0xFB, 0x02, 0x07, 0x00];
    await handlePackageDataWrite(cmd);
  }

  void _handleBatteryStatusResponse(List<int> data) {
    double battery = data.elementAt(6).toDouble();
    CapnoTrainerDataPoint batData = CapnoTrainerDataPoint(samplesCo2 * 1000 / rateCo2, battery);
    _onDataReceived([batData], CapnoTraienrStatusCodes.CODE_BATTERY_DATA);
    if (_printLog) print ("[CapnoTrainer SDK] CapnoTrainer GO Battery is $battery %");
    _isReceivedBattery = true;
  }

}

examples/lib/widgets.dart
import 'dart:io';
import 'package:intl/intl.dart';
import 'package:capnotrainer_sdk/capnotrainer_utils.dart';
import 'package:flutter/material.dart';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:capnotrainer_sdk/capnotrainer_sdk.dart';

class BLEScreen extends StatefulWidget {
  BLEScreen() : super();
  final List<BleCapnos> capnoDevices = <BleCapnos>[];
  final List<BluetoothDevice> bleDevices = <BluetoothDevice>[];
  String bleStatusMsg = "Status: None";
  String btnText = "Connect";

  List<CapnoTrainerDataPoint> co2 = [];
  double etco2 = 0;
  double battery = 0;
  double bpm = 0;
  double insp_co2 = 0;
  bool airwayBlocked = false;

  @override
  _BLEScreenState createState() => _BLEScreenState();
}

class _BLEScreenState extends State<BLEScreen> {
  CapnoTrainer capnoTrainer = CapnoTrainer();

  _addDeviceTolist(final BluetoothDevice device) async {

    if (!widget.bleDevices.contains(device)){
      setState(() {
        widget.bleDevices.add(device);
        widget.capnoDevices.add( BleCapnos(device) );
      });
    }

  }

  void handleCheckBluetoothAndLocationPermissions() async {
    Permission locationPermission = Permission.locationWhenInUse;
    Permission bluetoothPermission = Permission.bluetoothScan;

    PermissionStatus locationStatus = await  locationPermission.status;
    PermissionStatus bluetoothStatus = await  bluetoothPermission.status;

    if ( locationStatus != PermissionStatus.granted ) {
      await locationPermission.request();
    }
    if (Platform.isAndroid){
      if ( bluetoothStatus != PermissionStatus.granted ) {
        await bluetoothPermission.request();
      }
    }
  }

  Future<bool> _checkDeviceLocationIsOn() async {
    return await Permission.locationWhenInUse.serviceStatus.isEnabled;
  }

  Future<bool> _checkDeviceBluetoothIsOn() async {
    return await FlutterBluePlus.adapterState.first == BluetoothAdapterState.on;
  }

  void handleBluetoothScans() async {

    bool isLocation = await _checkDeviceLocationIsOn();
    bool isBle = await _checkDeviceBluetoothIsOn();

    if (!isLocation){ setState( () =>   widget.bleStatusMsg = "Status: Turn on mobile location");}
    if (!isBle){ setState(() => widget.bleStatusMsg = "Status: Turn on mobile bluetooth adapter");}
    if ( !isBle || !isLocation) { return ; }
    setState(() =>   widget.bleStatusMsg = "Status: Scanning..." );
    setState(() =>  widget.capnoDevices.length = 0);
    setState(() =>  widget.bleDevices.length = 0);

    Guid goGoServiceGuid = Guid("00001000-0000-1000-8000-00805f9b34fb");
    List<Guid> scanServices = [goGoServiceGuid];

    try {
      FlutterBluePlus.isSupported.then((isSupported) {
        if (isSupported) FlutterBluePlus.setLogLevel(LogLevel.info);
      });

      FlutterBluePlus.adapterState.listen((event) {
        setState(() {});
      });

      FlutterBluePlus.connectedDevices.forEach((element) {
        _addDeviceTolist(element);
      });

      FlutterBluePlus.scanResults.listen((List<ScanResult> results) {
        for (ScanResult result in results) {
          _addDeviceTolist(result.device);
        }
      });

      FlutterBluePlus.startScan(withServices: scanServices);
      // FlutterBluePlus.startScan();
    } catch (e) {
      print(e.toString());
    }
  }

  void onDataReceived(List<CapnoTrainerDataPoint> data, CapnoTraienrStatusCodes code){

    switch(code){
      case CapnoTraienrStatusCodes.CODE_CONNECTED:
        {
          setState(() => widget.bleStatusMsg = "Status: Connected to ${capnoTrainer.device.advName}");
          setState(() => widget.btnText = "Disconnect");
        } break;
      case CapnoTraienrStatusCodes.CODE_DISCONNECTED:
        {
          setState(() {
            widget.btnText = "Connect";
            widget.battery = 0;
            widget.etco2 = 0;
            widget.bpm = 0;
          });
          handleBluetoothScans();
        } break;
      case CapnoTraienrStatusCodes.CODE_RAW_CO2_DATA:
        {
          setState(() {
            widget.co2.addAll(data);
            if (widget.co2.length > 10){
             widget.co2.removeRange(0, widget.co2.length - 10);
            }
          });
        } break;
      case CapnoTraienrStatusCodes.CODE_BATTERY_DATA:
        {
          setState(() => widget.battery = data.last.y );
        } break;
      case CapnoTraienrStatusCodes.CODE_ETCO2_DATA:
        {
          setState(() => widget.etco2 = data.last.y );
        } break;
      case CapnoTraienrStatusCodes.CODE_BPM_DATA:
        {
          setState(() => widget.bpm = data.last.y );
        } break;
      case CapnoTraienrStatusCodes.CODE_INSP_CO2_DATA:
        {
          setState(() => widget.insp_co2 = data.last.y );
        } break;
      case CapnoTraienrStatusCodes.CODE_AIRWAY_BLOCKED:
        {
          setState(() => widget.airwayBlocked = true );
        } break;
      default: {

      }
    }
  }

  void handleButton(int index) async{
    if (!capnoTrainer.isConnected){
      setState(() => widget.bleStatusMsg = "Status: Connecting to ${widget.bleDevices.elementAt(index).advName}");
      await FlutterBluePlus.stopScan();
      await capnoTrainer.connect(widget.bleDevices.elementAt(index), onDataReceived);
    } else {
      await capnoTrainer.disconnect();
    }
    print (capnoTrainer.isConnected);
  }

  @override
  void initState() {
    super.initState();
    handleCheckBluetoothAndLocationPermissions();
    handleBluetoothScans();
  }

  @override
  void dispose() {
    FlutterBluePlus.stopScan();
    // TODO: implement dispose
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => _bleScreenView();

  Widget _bleScreenView() {
    return Container(
      height: MediaQuery.of(context).size.height,
      width: MediaQuery.of(context).size.width,
      padding: const EdgeInsets.only(left: 20, right: 20),
      child: SingleChildScrollView(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(widget.bleStatusMsg),
            const Divider(
              thickness: 1,
            ),
            _buildDeviceView()
          ],
        ),
      ),
    );
  }

  Widget _buildDeviceView() {
    return Column(
      children: [
        SizedBox(
          height: MediaQuery.of(context).size.height * .75,
          child: ListView.builder(
            itemCount: widget.capnoDevices.length,
            itemBuilder: (context, index) => Column(
              children: [
                ListTile(
                  contentPadding: EdgeInsets.fromLTRB(5, 5, 5, 5),
                  minLeadingWidth: 0,
                  horizontalTitleGap: 4,
                  leading: const Icon(
                    Icons.bluetooth,
                    size: 36,
                  ),
                  title: Text(
                    widget.capnoDevices.elementAt(index).title,
                    style: const TextStyle(
                        fontWeight: FontWeight.bold, fontSize: 16),
                  ),
                  subtitle: Text(
                    widget.capnoDevices.elementAt(index).subtitle,
                    style: const TextStyle(
                        fontWeight: FontWeight.w300, fontSize: 13),
                  ),
                  trailing: OutlinedButton(
                    onPressed: () async {handleButton(index);},
                    child: Text(widget.btnText),
                  )),
                  
                const Divider(thickness: 2,),
                Text("Battery: ${widget.battery}%  |  Airway Blocked: ${widget.airwayBlocked ? 'Yes' : 'No'}"),
                const Divider(thickness: 2,),
                Text("InspCO2: ${widget.insp_co2} mmHg  | "
                    " BPM:  ${widget.bpm} | "
                    " ETCO2: ${widget.etco2} mmHg"
                ),

                const Divider(thickness: 2,),
                const Text("Raw CO2 Data"),
                SizedBox(height: 5,),
                SizedBox(
                  height: MediaQuery.of(context).size.height*0.3,
                  width: MediaQuery.of(context).size.width,
                  child: _tableView(context),
                )
              ],
            ))),
      ],
    );
  }

  Widget _tableView(BuildContext context) {
    return ListView.builder(
      itemCount: widget.co2.length,
      itemBuilder: (context, index) {
        return Row(
          children: <Widget>[
            Expanded(
              child: Container(
                padding: EdgeInsets.all(8.0),
                decoration: BoxDecoration(
                  border: Border.all(color: Colors.grey),
                ),
                child: Text(
                  DateFormat("mm:ss")
                      .format(
                        DateTime.fromMillisecondsSinceEpoch(
                            widget.co2[index].x.toInt()
                        )
                  ).toString()
                ),
              ),
            ),
            Expanded(
              child: Container(
                padding: EdgeInsets.all(8.0),
                decoration: BoxDecoration(
                  border: Border.all(color: Colors.grey),
                ),
                child: Text("${widget.co2[index].y.toString()} mmHg"),
              ),
            ),
          ],
        );
      },
    );
  }

}

class BleCapnos {
  BluetoothDevice capno;
  String title = "";
  String subtitle = "";
  BleCapnos(this.capno){
    this.title = this.capno.advName;
    this.subtitle = "CapnoTrainer GO";
  }
}
